### ISSUE TITLE
Feature: Implement Genome Encoding Model with Autoencoder Compression for Agents

### ISSUE THREAD

**Issue by csmangum** [Dooders/Experiments#34]
This feature aims to enhance the current genome encoding model by integrating an **autoencoder** to compress and encode neural network weights into a latent vector. The autoencoder will replace the direct storage of network weights in the genome, resulting in a more compact genome representation. This compressed representation will be used during agent creation, mutation, and crossover processes.

---

### Objectives

1. **Train an Autoencoder**:
   - Develop and train an autoencoder to compress neural network weights into a latent vector.
   - Ensure accurate reconstruction of weights with minimal loss.

2. **Integrate the Autoencoder**:
   - Replace the direct weight encoding in the genome with the autoencoder's latent vector.
   - Modify the `to_genome` and `from_genome` methods to use the autoencoder for encoding and decoding weights.

3. **Update Evolutionary Processes**:
   - Update mutation and crossover functions to handle the latent vector representation.
   - Ensure genetic operations remain efficient and compatible with the compressed genome.

4. **Validate and Test**:
   - Test the updated genome encoding model to confirm that agents can be accurately reconstructed and evolved.
   - Compare the performance of the new model with the previous weight-encoding approach.

---

#### 1. **Autoencoder Development**
   - Define and implement the autoencoder architecture:
     - Input size: Total number of neural network weights.
     - Latent size: Compressed dimension for the latent vector.
   - Prepare a dataset of neural network weights from diverse agents.
   - Train the autoencoder using a reconstruction loss function (e.g., MSE).
   - Save the trained autoencoder for use during genome encoding.

#### 2. **Modify Genome Representation**
   - Update the `to_genome` method:
     - Extract and flatten neural network weights.
     - Pass weights through the trained autoencoder encoder to get the latent vector.
     - Replace weight encoding in the genome with the latent vector.
   - Update the `from_genome` method:
     - Decode the latent vector using the trained decoder.
     - Reconstruct the neural network weights from the latent vector.
     - Assign the reconstructed weights to the agent's neural network modules.

#### 3. **Update Genetic Operators**
   - **Mutation**:
     - Apply Gaussian mutation to the latent vector in the genome.
     - Ensure the mutation scale is appropriate for the latent space.
   - **Crossover**:
     - Perform crossover operations (e.g., single-point, uniform) on the latent vectors.
     - Ensure the latent vector remains within a valid range after crossover.

#### 4. **Validation and Testing**
   - Verify the autoencoder accurately compresses and reconstructs neural network weights.
   - Test the evolutionary process with the new genome representation:
     - Validate agent reconstruction from genome.
     - Observe agent performance and fitness evolution.
   - Compare with the previous implementation:
     - Measure genome size reduction.
     - Analyze performance improvements in evolution.

---

### Acceptance Criteria

- [ ] Autoencoder successfully trained and saved for encoding/decoding neural network weights.
- [ ] `to_genome` and `from_genome` methods updated to use the autoencoder for weight compression.
- [ ] Mutation and crossover functions updated to handle the latent vector.
- [ ] Agents can be accurately reconstructed and evolved using the new genome representation.
- [ ] Performance comparisons between the old and new genome encoding models documented.

---

### Additional Notes

- **Autoencoder Training**:
  - Experiment with different latent vector sizes to find a balance between compression and reconstruction accuracy.
  - Consider using diverse agents for training to improve generalizability.

- **Integration**:
  - Ensure the autoencoder's decoder remains fixed during the evolutionary process to maintain consistency.

- **Optimization**:
  - Profile the system for any performance bottlenecks introduced by the autoencoder.

**Comment by csmangum** [comment]
Using an autoencoder allows you to compress complex data like neural network weights into a lower-dimensional latent space. This compressed representation can then be used as part of your genome, reducing the genome size and potentially improving the efficiency of your evolutionary algorithm.


1. **What an Autoencoder Is**: A brief overview of autoencoders.
2. **Why Use an Autoencoder in Your Genome System**: Benefits and considerations.
3. **How to Integrate an Autoencoder**: Steps to include or replace parts of your genome system with an autoencoder.
4. **Implementation Details**: Code examples and guidance on implementing this approach.
5. **Potential Challenges and Solutions**: Addressing issues you might encounter.

---

### 1. What an Autoencoder Is

An **autoencoder** is a type of neural network designed to learn a compressed representation (encoding) of input data, typically for the purpose of dimensionality reduction. It consists of two parts:

- **Encoder**: Compresses the input data into a lower-dimensional latent space.
- **Decoder**: Reconstructs the original data from the compressed representation.

Autoencoders are trained to minimize the difference between the input and the reconstructed output, effectively learning the most important features of the data.

---

### 2. Why Use an Autoencoder in Your Genome System

**Benefits:**

- **Reduced Genome Size**: Compressing neural network weights results in a smaller genome, making evolutionary operations more efficient.
- **Improved Evolutionary Search**: Operating in a lower-dimensional latent space can help the evolutionary algorithm explore the solution space more effectively.
- **Abstract Representation**: The latent space captures essential features, potentially leading to more generalizable agents.

**Considerations:**

- **Training the Autoencoder**: You'll need to train the autoencoder separately to ensure it can accurately compress and decompress the neural network weights.
- **Fixed Decoder**: In the evolutionary process, you typically keep the decoder fixed to ensure consistency when reconstructing agents from genomes.
- **Complexity**: Introducing an autoencoder adds complexity to your system, requiring careful integration.

---

### 3. How to Integrate an Autoencoder

#### **Step 1: Collect Neural Network Weights**

- Gather a dataset of neural network weights from existing agents.
- This dataset will be used to train the autoencoder.

#### **Step 2: Train the Autoencoder**

- **Encoder**: Compresses the neural network weights into a lower-dimensional latent vector.
- **Decoder**: Reconstructs the neural network weights from the latent vector.
- Train the autoencoder to minimize the reconstruction error.

#### **Step 3: Replace Direct Weights Encoding with Latent Vectors**

- Instead of encoding the full neural network weights in the genome, encode the latent vector produced by the encoder.
- The genome now includes:
  - **Latent Vector**: Compressed representation of neural network weights.
  - **Other Parameters**: Action weights, resource levels, health, etc.

#### **Step 4: Decoding During Agent Reconstruction**

- When creating an agent from a genome, use the decoder to reconstruct the neural network weights from the latent vector.
- Assign the reconstructed weights to the agent's neural network modules.

#### **Step 5: Evolutionary Operations on Latent Vectors**

- **Mutation**: Apply mutations to the latent vector in the genome.
- **Crossover**: Combine latent vectors from parent genomes during crossover.
- Since the latent space is continuous and lower-dimensional, these operations can be more efficient.

---

### 4. Implementation Details

#### **a. Training the Autoencoder**

**Data Preparation:**

- Collect neural network weights from a diverse set of agents.
- Flatten the weights into a single vector for each agent.

**Autoencoder Architecture:**

- **Input Layer**: Size equal to the total number of weights.
- **Encoder Layers**: Several hidden layers reducing in size.
- **Latent Layer**: Represents the compressed latent vector.
- **Decoder Layers**: Mirror of the encoder layers, expanding back to the original size.

**Training Procedure:**

- Use a reconstruction loss function, such as Mean Squared Error (MSE), to measure the difference between the original and reconstructed weights.
- Train the autoencoder until the reconstruction error is acceptably low.

**Example Code:**

```python
import torch
import torch.nn as nn
import torch.optim as optim

# Define the autoencoder model
class WeightAutoencoder(nn.Module):
    def __init__(self, input_size, latent_size):
        super(WeightAutoencoder, self).__init__()
        self.encoder = nn.Sequential(
            nn.Linear(input_size, 512),
            nn.ReLU(),
            nn.Linear(512, latent_size)
        )
        self.decoder = nn.Sequential(
            nn.Linear(latent_size, 512),
            nn.ReLU(),
            nn.Linear(512, input_size)
        )

    def forward(self, x):
        latent = self.encoder(x)
        reconstructed = self.decoder(latent)
        return reconstructed, latent
```

**Training Loop:**

```python
# Assume 'weights_dataset' is a PyTorch DataLoader providing weight vectors
input_size = total_number_of_weights  # From your agent's neural networks
latent_size = 50  # Choose based on desired compression

autoencoder = WeightAutoencoder(input_size, latent_size)
criterion = nn.MSELoss()
optimizer = optim.Adam(autoencoder.parameters(), lr=0.001)

num_epochs = 50
for epoch in range(num_epochs):
    for data in weights_dataset:
        optimizer.zero_grad()
        reconstructed, _ = autoencoder(data)
        loss = criterion(reconstructed, data)
        loss.backward()
        optimizer.step()
    print(f"Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}")
```

#### **b. Modifying the Genome Structure**

**Updated Genome:**

```python
def to_genome(self) -> dict:
    # Get flattened neural network weights
    weights = []
    for module_name in dir(self):
        if module_name.endswith('_module'):
            module = getattr(self, module_name)
            if hasattr(module, 'state_dict'):
                state_dict = module.state_dict()
                for tensor in state_dict.values():
                    weights.extend(tensor.flatten().tolist())

    # Convert weights to a tensor
    weights_tensor = torch.tensor(weights, dtype=torch.float32)
    # Pass through encoder to get latent vector
    with torch.no_grad():
        latent_vector = autoencoder.encoder(weights_tensor).numpy()

    genome = {
        'latent_vector': latent_vector.tolist(),
        'action_weights': [action.weight for action in self.actions],
        'resource_level': self.resource_level / self.config.max_resource_level,
        'current_health': self.current_health / self.max_health,
        # Include other parameters as needed
    }
    return genome
```

#### **c. Reconstructing Agents from the Genome**

**Updated `from_genome` Method:**

```python
@classmethod
def from_genome(
    cls,
    genome: dict,
    agent_id: int,
    position: tuple[int, int],
    environment: "Environment",
) -> "BaseAgent":
    # Reconstruct neural network weights from latent vector
    latent_vector = torch.tensor(genome['latent_vector'], dtype=torch.float32)
    with torch.no_grad():
        reconstructed_weights = autoencoder.decoder(latent_vector).numpy()

    # Assign weights to neural network modules
    idx = 0
    for module_name in dir(cls):
        if module_name.endswith('_module'):
            module = getattr(cls, module_name)
            if hasattr(module, 'state_dict'):
                state_dict = module.state_dict()
                new_state_dict = {}
                for key, param in state_dict.items():
                    param_size = param.numel()
                    param_shape = param.shape
                    param_data = reconstructed_weights[idx:idx + param_size]
                    param_tensor = torch.tensor(param_data).view(param_shape)
                    new_state_dict[key] = param_tensor
                    idx += param_size
                module.load_state_dict(new_state_dict)

    # Create new agent instance
    agent = cls(
        agent_id=agent_id,
        position=position,
        resource_level=genome['resource_level'] * cls.config.max_resource_level,
        environment=environment,
        action_set=[
            Action(name, weight, globals()[f"{name}_action"])
            for name, weight in zip(
                [action.name for action in BASE_ACTION_SET],
                genome['action_weights']
            )
        ],
    )

    # Set health and other parameters
    agent.current_health = genome['current_health'] * agent.max_health
    # Assign reconstructed modules
    for module_name in dir(agent):
        if module_name.endswith('_module') and hasattr(cls, module_name):
            setattr(agent, module_name, getattr(cls, module_name))

    return agent
```

#### **d. Evolutionary Operations on Latent Vectors**

**Mutation:**

```python
def mutate_genome(genome, mutation_rate=0.01):
    # Mutate latent vector
    genome['latent_vector'] = [
        value + random.gauss(0, mutation_rate) for value in genome['latent_vector']
    ]

    # Mutate other parameters as before
    # ...

    return genome
```

**Crossover:**

```python
def crossover_genomes(genome1, genome2):
    child_genome = {}

    # Crossover latent vectors (e.g., single-point crossover)
    crossover_point = random.randint(1, len(genome1['latent_vector']) - 1)
    child_genome['latent_vector'] = (
        genome1['latent_vector'][:crossover_point] +
        genome2['latent_vector'][crossover_point:]
    )

    # Crossover other parameters
    # ...

    return child_genome
```

---

### 5. Potential Challenges and Solutions

#### **a. Training the Autoencoder**

**Challenge:** Ensuring the autoencoder accurately reconstructs neural network weights.

**Solution:**

- Use a sufficiently large and diverse dataset of weights.
- Experiment with different architectures and hyperparameters.
- Consider using regularization techniques to prevent overfitting.

#### **b. Compatibility Between Encoder and Decoder**

**Challenge:** The decoder must remain consistent during the evolutionary process.

**Solution:**

- Once trained, keep the autoencoder's weights fixed during evolution.
- If you need to update the autoencoder, ensure compatibility or retrain agents accordingly.

#### **c. Latent Space Interpretability**

**Challenge:** Latent vectors may not correspond directly to meaningful agent behaviors.

**Solution:**

- While the latent space may be abstract, evolutionary processes can still optimize agents based on fitness.
- Monitor and analyze the latent space to understand its structure and impact on agent performance.

#### **d. Integration Complexity**

**Challenge:** Incorporating an autoencoder adds complexity to your system.

**Solution:**

- Modularize your code to separate concerns (e.g., autoencoder training, agent evolution).
- Thoroughly test each component individually before integration.
- Document the workflow and dependencies.

---

### **Conclusion**

Including an autoencoder in your genome system can effectively reduce the size of the genome by compressing neural network weights into a latent vector. This approach allows you to:

- **Efficiently Evolve Agents**: Smaller genomes make genetic operations more manageable.
- **Explore Compressed Representations**: The latent space may capture essential features leading to more robust agents.

**Next Steps:**

1. **Train the Autoencoder**: Use existing agent weights to train the autoencoder until reconstruction loss is minimized.
2. **Update Genome Representation**: Modify your `to_genome` and `from_genome` methods to use the latent vector.
3. **Adjust Evolutionary Algorithms**: Ensure mutation and crossover operations are compatible with the latent vector.
4. **Test and Validate**: Run experiments to validate that agents can be accurately reconstructed and evolved using this method.
5. **Monitor Performance**: Compare the performance and evolution speed with and without the autoencoder to assess benefits.

---

**Additional Resources**

- **Autoencoder Tutorials**:
  - [PyTorch Tutorial on Autoencoders](https://github.com/pytorch/examples/tree/master/vae)
  - [Introduction to Autoencoders](https://blog.keras.io/building-autoencoders-in-keras.html)

- **Research Papers**:
  - *Deep Neuroevolution: Genetic Algorithms Are a Competitive Alternative for Training Deep Neural Networks for Reinforcement Learning* by Such et al.

- **Libraries**:
  - **PyTorch**: For building and training neural networks.
  - **NumPy**: For numerical operations.
